content/docs/token-exchange.md

```diff
@@ -0,0 +1,212 @@
+---
+title: "Machine Authentication to Dex"
+linkTitle: "Authentication for Machines"
+description: ""
+date: 2023-07-01
+draft: false
+toc: true
+weight: 1080
+---
+
+## Overview
+
+Most Dex connectors redirect the user to the upstream identity provider as part of the authentication flow.
+While this works for human users,
+it is much harder for machines and automated processes (e.g. CI pipelines) to complete this interactive flow.
+This is where [OAuth2 Token Exchange][token-exchange] comes in:
+it allows clients to exchange an access or id token they already have
+(obtained from their environment, though custom cli commands, etc.)
+for a token issued by dex.
+
+This works like [GCP Workload Identity Federation][gcp-federation] and [AWS Web Identity Federation][aws-federation],
+allowing processes running in trusted execution environments that issue OIDC tokens,
+such as [Gtihub Actions][gh-actions], [Buildkite][buildkite], [CircleCI][circleci], [GCP][gcp], and others,
+to exchange them for a dex issued token to access protected resources.
+
+The authentication flow looks like:
+
+1. Client independently obtains an access / id token from the upstream IDP.
+2. Client exchanges the upstream token for a dex access / id token via the token exchange flow.
+3. Use token to access dex protected resources.
+4. Repeat from step 1 when the token expires.
+
+## Configuring dex
+
+Currently, only the [OIDC Connector][oidc-connector] supports token exchanges.
+For this flow, `clientID`, `clientSecret`, and `redirectURI` aren't required.
+
+```yaml
+issuer: https://dex.example.com
+storage:
+    type: sqlite3
+    config:
+        file: dex.db
+web:
+    http: 0.0.0.0:8001
+
+outh2:
+  grantTypes:
+    # ensure grantTypes includes the token-exchange grant (default)
+    - "urn:ietf:params:oauth:grant-type:token-exchange"
+
+connectors:
+  - name: My Upstream
+    type: oidc
+    id: my-upstream
+    config:
+      # The client submitted subject token will be verified against the issuer given here.
+      issuer: https://token.example.com
+      # Additional scopes in token response, supported list at:
+      # https://dexidp.io/docs/custom-scopes-claims-clients/#scopes
+      scopes:
+        - groups
+        - federated:id
+      # mapping of fields from the submitted token
+      userNameKey: sub
+
+staticClients:
+  # dex issued tokens are bound to clients.
+  # For the token exchange flow, the client id and secret pair must be submitted as the username:password
+  # via Basic Authentication.
+  - name: My App
+    id: my-app
+    secret: my-secret
+    # We set public to indicate we don't intend to keep the client secret actually secret.
+    # https://dexidp.io/docs/custom-scopes-claims-clients/#public-clients
+    public: true
```

Comment:
I'm confused by the presence of the `staticClients` block -- when I first read this, my mental model was that it would be necessary to declare clients (either in a `staticClients` block or via the gRPC API) in order to conduct token exchange for them. However, in the example below that shows how token exchange can be used in Github Actions (which happens to be something I am working on right now, so this doc update is quite timely for me!) there are no `staticClients` configured in the Dex config, nor does there appear to be any interaction with Dex aside from the token exchange POST to the `/token` endpoint. Could you clarify whether or not a client would need to be configured (through `staticClients` or otherwise) in order to support token exchange for a given client? 

It seems that because the client ID and secret must be provided in the basic auth for the token exchange post, that perhaps Dex has all the information it needs without the client actually having been configured, but in that case I do not understand why the client is configured here, since this config seems to only allow token exchange. I understand it would be necessary to include the client when we want to implement both token exchange and user-mediated auth flows, and perhaps the purpose of including this block in the example is so you can include the comment regarding setting `public: true` because the existence of the token exchange flow indicates that the client secret is not really a secret, but at this point I feel I am guessing quite a lot and it would be very helpful for the documentation to clarify this.

Reply:
I agree that explaining connectors/clients should be out of scope for this page. I think my main confusion came from the missing config block, as I originally had the right idea that a (public) client needs to be configured.

I do think that since the use of token exchange necessitates treating the client as public as well as the use-cases involving token exchange being likely to be apps that also have real (i.e. human) users, it would be helpful to include the example (essentially from your comment) of how one might want to employ cross-client trust to trust the public client and then request the scope of the main application client. Or at a minimum, these docs could mention that with a reference to the main [cross client trust](https://dexidp.io/docs/custom-scopes-claims-clients/#cross-client-trust-and-authorized-party) docs without giving an example (though I would note that those docs do not _seem_ to specifically address the use-case of trusting a public client, though perhaps the section on [public clients](https://dexidp.io/docs/custom-scopes-claims-clients/#public-clients) would be a more appropriate place to include that as an option, and this doc could then just reference that).

Reply:
Thanks, that looks great.